File: packages/sdk-go/circuit_breaker_test.go
package rollgate

import (
	"errors"
	"testing"
	"time"
)

func TestCircuitBreaker_InitialState(t *testing.T) {
	cb := NewCircuitBreaker(DefaultCircuitBreakerConfig())

	if cb.GetState() != CircuitStateClosed {
		t.Errorf("expected initial state to be closed, got %s", cb.GetState())
	}

	if !cb.IsAllowingRequests() {
		t.Error("expected to allow requests initially")
	}

	stats := cb.GetStats()
	if stats.Failures != 0 {
		t.Errorf("expected 0 failures, got %d", stats.Failures)
	}
}

func TestCircuitBreaker_Execute(t *testing.T) {
	t.Run("should pass through successful requests", func(t *testing.T) {
		cb := NewCircuitBreaker(DefaultCircuitBreakerConfig())
		called := false

		err := cb.Execute(func() error {
			called = true
			return nil
		})

		if err != nil {
			t.Errorf("expected no error, got %v", err)
		}
		if !called {
			t.Error("expected function to be called")
		}
		if cb.GetState() != CircuitStateClosed {
			t.Error("expected state to remain closed")
		}
	})

	t.Run("should pass through and rethrow errors", func(t *testing.T) {
		cb := NewCircuitBreaker(DefaultCircuitBreakerConfig())
		expectedErr := errors.New("test error")

		err := cb.Execute(func() error {
			return expectedErr
		})

		if err != expectedErr {
			t.Errorf("expected error %v, got %v", expectedErr, err)
		}
	})

	t.Run("should track failures", func(t *testing.T) {
		cb := NewCircuitBreaker(DefaultCircuitBreakerConfig())

		_ = cb.Execute(func() error {
			return errors.New("failure")
		})

		stats := cb.GetStats()
		if stats.Failures != 1 {
			t.Errorf("expected 1 failure, got %d", stats.Failures)
		}
	})
}

func TestCircuitBreaker_Opening(t *testing.T) {
	t.Run("should open after reaching failure threshold", func(t *testing.T) {
		config := DefaultCircuitBreakerConfig()
		config.FailureThreshold = 3
		cb := NewCircuitBreaker(config)

		for i := 0; i < 3; i++ {
			_ = cb.Execute(func() error {
				return errors.New("failure")
			})
		}

		if cb.GetState() != CircuitStateOpen {
			t.Errorf("expected state to be open, got %s", cb.GetState())
		}
	})

	t.Run("should throw CircuitOpenError when open", func(t *testing.T) {
		config := DefaultCircuitBreakerConfig()
		config.FailureThreshold = 1
		config.RecoveryTimeout = 1 * time.Hour
		cb := NewCircuitBreaker(config)

		_ = cb.Execute(func() error {
			return errors.New("failure")
		})

		if cb.GetState() != CircuitStateOpen {
			t.Errorf("expected state to be open, got %s", cb.GetState())
		}

		err := cb.Execute(func() error {
			return nil
		})

		if err != ErrCircuitOpen {
			t.Errorf("expected ErrCircuitOpen, got %v", err)
		}
	})
}

func TestCircuitBreaker_MonitoringWindow(t *testing.T) {
	config := DefaultCircuitBreakerConfig()
	config.FailureThreshold = 3
	config.MonitoringWindow = 50 * time.Millisecond
	cb := NewCircuitBreaker(config)

	// First failure
	_ = cb.Execute(func() error {
		return errors.New("failure")
	})

	stats := cb.GetStats()
	if stats.Failures != 1 {
		t.Errorf("expected 1 failure, got %d", stats.Failures)
	}

	// Wait for monitoring window to expire
	time.Sleep(60 * time.Millisecond)

	// Second failure (first should be expired)
	_ = cb.Execute(func() error {
		return errors.New("failure")
	})

	stats = cb.GetStats()
	if stats.Failures != 1 {
		t.Errorf("expected 1 failure after window expiry, got %d", stats.Failures)
	}

	// Circuit should still be closed
	if cb.GetState() != CircuitStateClosed {
		t.Error("expected state to remain closed")
	}
}

func TestCircuitBreaker_Recovery(t *testing.T) {
	t.Run("should transition to half-open after recovery timeout", func(t *testing.T) {
		config := DefaultCircuitBreakerConfig()
		config.FailureThreshold = 1
		config.RecoveryTimeout = 20 * time.Millisecond
		cb := NewCircuitBreaker(config)

		_ = cb.Execute(func() error {
			return errors.New("failure")
		})

		if cb.GetState() != CircuitStateOpen {
			t.Error("expected state to be open")
		}

		// Wait for recovery timeout
		time.Sleep(25 * time.Millisecond)

		if !cb.IsAllowingRequests() {
			t.Error("expected to allow requests after recovery timeout")
		}
	})

	t.Run("should close circuit after success in half-open", func(t *testing.T) {
		config := DefaultCircuitBreakerConfig()
		config.FailureThreshold = 1
		config.RecoveryTimeout = 10 * time.Millisecond
		config.SuccessThreshold = 1
		cb := NewCircuitBreaker(config)

		_ = cb.Execute(func() error {
			return errors.New("failure")
		})

		time.Sleep(15 * time.Millisecond)

		err := cb.Execute(func() error {
			return nil
		})

		if err != nil {
			t.Errorf("expected no error, got %v", err)
		}
		if cb.GetState() != CircuitStateClosed {
			t.Errorf("expected state to be closed, got %s", cb.GetState())
		}
	})

	t.Run("should reopen on failure in half-open", func(t *testing.T) {
		config := DefaultCircuitBreakerConfig()
		config.FailureThreshold = 1
		config.RecoveryTimeout = 10 * time.Millisecond
		config.SuccessThreshold = 3
		cb := NewCircuitBreaker(config)

		_ = cb.Execute(func() error {
			return errors.New("failure")
		})

		time.Sleep(15 * time.Millisecond)

		_ = cb.Execute(func() error {
			return errors.New("failure again")
		})

		if cb.GetState() != CircuitStateOpen {
			t.Errorf("expected state to be open, got %s", cb.GetState())
		}
	})
}

func TestCircuitBreaker_ForceReset(t *testing.T) {
	config := DefaultCircuitBreakerConfig()
	config.FailureThreshold = 1
	cb := NewCircuitBreaker(config)

	_ = cb.Execute(func() error {
		return errors.New("failure")
	})

	if cb.GetState() != CircuitStateOpen {
		t.Error("expected state to be open")
	}

	cb.ForceReset()

	if cb.GetState() != CircuitStateClosed {
		t.Errorf("expected state to be closed after reset, got %s", cb.GetState())
	}

	stats := cb.GetStats()
	if stats.Failures != 0 {
		t.Errorf("expected 0 failures after reset, got %d", stats.Failures)
	}
}

func TestCircuitBreaker_ForceOpen(t *testing.T) {
	cb := NewCircuitBreaker(DefaultCircuitBreakerConfig())

	cb.ForceOpen()

	if cb.GetState() != CircuitStateOpen {
		t.Errorf("expected state to be open, got %s", cb.GetState())
	}
}

func TestCircuitBreaker_StateChangeCallback(t *testing.T) {
	config := DefaultCircuitBreakerConfig()
	config.FailureThreshold = 1
	cb := NewCircuitBreaker(config)

	var fromState, toState CircuitState
	cb.OnStateChange(func(from, to CircuitState) {
		fromState = from
		toState = to
	})

	_ = cb.Execute(func() error {
		return errors.New("failure")
	})

	if fromState != CircuitStateClosed {
		t.Errorf("expected from state to be closed, got %s", fromState)
	}
	if toState != CircuitStateOpen {
		t.Errorf("expected to state to be open, got %s", toState)
	}
}

File: packages/sdk-core/src/circuit-breaker.test.ts
import { describe, it, expect, beforeEach, afterEach, vi } from "vitest";
import {
  CircuitBreaker,
  CircuitState,
  CircuitOpenError,
  DEFAULT_CIRCUIT_BREAKER_CONFIG,
} from "./circuit-breaker";

describe("CircuitBreaker", () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  describe("DEFAULT_CIRCUIT_BREAKER_CONFIG", () => {
    it("should have sensible defaults", () => {
      expect(DEFAULT_CIRCUIT_BREAKER_CONFIG.failureThreshold).toBe(5);
      expect(DEFAULT_CIRCUIT_BREAKER_CONFIG.recoveryTimeout).toBe(30000);
      expect(DEFAULT_CIRCUIT_BREAKER_CONFIG.monitoringWindow).toBe(60000);
      expect(DEFAULT_CIRCUIT_BREAKER_CONFIG.successThreshold).toBe(3);
    });
  });

  describe("initial state", () => {
    it("should start in CLOSED state", () => {
      const cb = new CircuitBreaker();
      expect(cb.getState()).toBe(CircuitState.CLOSED);
    });

    it("should allow requests initially", () => {
      const cb = new CircuitBreaker();
      expect(cb.isAllowingRequests()).toBe(true);
    });

    it("should have zero failures initially", () => {
      const cb = new CircuitBreaker();
      expect(cb.getStats().failures).toBe(0);
    });
  });

  describe("execute", () => {
    it("should pass through successful requests", async () => {
      const cb = new CircuitBreaker();
      const fn = vi.fn().mockResolvedValue("success");

      const result = await cb.execute(fn);

      expect(result).toBe("success");
      expect(fn).toHaveBeenCalledTimes(1);
      expect(cb.getState()).toBe(CircuitState.CLOSED);
    });

    it("should pass through and rethrow errors", async () => {
      const cb = new CircuitBreaker();
      const error = new Error("test error");
      const fn = vi.fn().mockRejectedValue(error);

      await expect(cb.execute(fn)).rejects.toThrow("test error");
      expect(fn).toHaveBeenCalledTimes(1);
    });

    it("should track failures", async () => {
      const cb = new CircuitBreaker();
      const fn = vi.fn().mockRejectedValue(new Error("failure"));

      await cb.execute(fn).catch(() => {});

      expect(cb.getStats().failures).toBe(1);
    });
  });

  describe("circuit opening", () => {
    it("should open after reaching failure threshold", async () => {
      const cb = new CircuitBreaker({ failureThreshold: 3 });
      const fn = vi.fn().mockRejectedValue(new Error("failure"));

      for (let i = 0; i < 3; i++) {
        await cb.execute(fn).catch(() => {});
      }

      expect(cb.getState()).toBe(CircuitState.OPEN);
    });

    it("should emit circuit-open event", async () => {
      const cb = new CircuitBreaker({ failureThreshold: 2 });
      const fn = vi.fn().mockRejectedValue(new Error("failure"));
      const openHandler = vi.fn();
      cb.on("circuit-open", openHandler);

      await cb.execute(fn).catch(() => {});
      await cb.execute(fn).catch(() => {});

      expect(openHandler).toHaveBeenCalledTimes(1);
      expect(openHandler).toHaveBeenCalledWith(
        expect.objectContaining({ failures: 2 }),
      );
    });

    it("should throw CircuitOpenError when open", async () => {
      const cb = new CircuitBreaker({
        failureThreshold: 1,
        recoveryTimeout: 30000,
      });
      const fn = vi.fn().mockRejectedValue(new Error("failure"));

      await cb.execute(fn).catch(() => {});

      expect(cb.getState()).toBe(CircuitState.OPEN);

      const successFn = vi.fn().mockResolvedValue("success");
      await expect(cb.execute(successFn)).rejects.toThrow(CircuitOpenError);
      expect(successFn).not.toHaveBeenCalled();
    });
  });

  describe("monitoring window", () => {
    it("should only count failures within the window", async () => {
      const cb = new CircuitBreaker({
        failureThreshold: 3,
        monitoringWindow: 5000,
      });
      const fn = vi.fn().mockRejectedValue(new Error("failure"));

      await cb.execute(fn).catch(() => {});
      expect(cb.getStats().failures).toBe(1);

      vi.advanceTimersByTime(6000);

      await cb.execute(fn).catch(() => {});
      expect(cb.getStats().failures).toBe(1);

      expect(cb.getState()).toBe(CircuitState.CLOSED);
    });
  });

  describe("recovery (half-open)", () => {
    it("should transition to HALF_OPEN after recovery timeout", async () => {
      const cb = new CircuitBreaker({
        failureThreshold: 1,
        recoveryTimeout: 5000,
      });
      const fn = vi.fn().mockRejectedValue(new Error("failure"));

      await cb.execute(fn).catch(() => {});
      expect(cb.getState()).toBe(CircuitState.OPEN);

      vi.advanceTimersByTime(5001);

      expect(cb.isAllowingRequests()).toBe(true);
    });

    it("should allow test request in HALF_OPEN state", async () => {
      const cb = new CircuitBreaker({
        failureThreshold: 1,
        recoveryTimeout: 1000,
        successThreshold: 1,
      });

      await cb.execute(vi.fn().mockRejectedValue(new Error())).catch(() => {});
      expect(cb.getState()).toBe(CircuitState.OPEN);

      vi.advanceTimersByTime(1001);

      const successFn = vi.fn().mockResolvedValue("success");
      const result = await cb.execute(successFn);

      expect(result).toBe("success");
      expect(cb.getState()).toBe(CircuitState.CLOSED);
    });

    it("should close circuit after success threshold in HALF_OPEN", async () => {
      const cb = new CircuitBreaker({
        failureThreshold: 1,
        recoveryTimeout: 1000,
        successThreshold: 2,
      });

      await cb.execute(vi.fn().mockRejectedValue(new Error())).catch(() => {});
      vi.advanceTimersByTime(1001);

      const successFn = vi.fn().mockResolvedValue("success");

      await cb.execute(successFn);
      expect(cb.getState()).toBe(CircuitState.HALF_OPEN);

      await cb.execute(successFn);
      expect(cb.getState()).toBe(CircuitState.CLOSED);
    });

    it("should reopen circuit on failure in HALF_OPEN", async () => {
      const cb = new CircuitBreaker({
        failureThreshold: 1,
        recoveryTimeout: 1000,
        successThreshold: 3,
      });

      await cb.execute(vi.fn().mockRejectedValue(new Error())).catch(() => {});
      vi.advanceTimersByTime(1001);

      await cb.execute(vi.fn().mockRejectedValue(new Error())).catch(() => {});

      expect(cb.getState()).toBe(CircuitState.OPEN);
    });
  });

  describe("forceReset", () => {
    it("should reset circuit to closed", async () => {
      const cb = new CircuitBreaker({ failureThreshold: 1 });

      await cb.execute(vi.fn().mockRejectedValue(new Error())).catch(() => {});
      expect(cb.getState()).toBe(CircuitState.OPEN);

      cb.forceReset();
      expect(cb.getState()).toBe(CircuitState.CLOSED);
      expect(cb.getStats().failures).toBe(0);
    });
  });

  describe("forceOpen", () => {
    it("should force circuit to open", () => {
      const cb = new CircuitBreaker();

      cb.forceOpen();
      expect(cb.getState()).toBe(CircuitState.OPEN);
    });
  });
});
